mitodl/ocw-studio
# use new mass build pipeline

# What are the relevant tickets?
Closes https://github.com/mitodl/ocw-studio/issues/1966

# Description (What does it do?)
This PR sets up `content_sync.pipelines.concourse.MassBuildSitesPipeline` to use `content_sync.pipelines.definitions.mass_build_sites.MassBuildSitesPipelineDefinition` rather than the current strategy of running `mass-build-sites.yml` through `get_pipeline_definition`, stripping lines out based on environment. It also makes a number of adjustments to tests relating to pipeline fixtures and the tests that use them. 


# How can this be tested?
 - Make sure you have a fairly recent database restore, have a Github organization set up and are otherwise set up for publishing sites locally
 - To facilitate easier local testing, you may also want to specifically set the following environment variables:
```
RESOURCE_BASE_URL_DRAFT=https://draft.ocw.mit.edu
RESOURCE_BASE_URL_LIVE=https://ocw.mit.edu
STATIC_API_BASE_URL_DRAFT=https://draft.ocw.mit.edu
STATIC_API_BASE_URL_LIVE=https://ocw.mit.edu
OCW_STUDIO_DRAFT_URL=http://localhost:8044/
OCW_STUDIO_LIVE_URL=http://localhost:8045/
OCW_HUGO_THEMES_BRANCH=main
OCW_HUGO_PROJECTS_BRANCH=main
```
 - You will also want to make sure you add `content_sync.tasks.update_mass_build_pipelines` to your `PREPUBLISH_ACTIONS` env variable, or create it if it doesn't exist
 - Make sure your Github repos in your test org are fully up to date by running the following (This will take a couple of hours depending on your latency to Github and rate limiting, so you may want to start this and then move on to something else for the time being):
```
docker compose exec web ./manage.py reset_sync_states --skip_sync
docker compose exec web ./manage.py mass_publish live
```
 - First we will test that the pipelines are upserted to Concourse when a site is being published for the first time to either draft or live:
   - If you don't have any mass build pipelines already in Concourse, you can skip the next 2 steps:
   - Run `fly -t local pipelines | grep mass-build-sites` and you should see some output listing the mass build pipelines in your local Concourse
   - Destroy the pipelines one at a time by running, for example, `fly -t local destroy-pipeline -p 'mass-build-sites/offline:false,prefix:"",projects_branch:main,starter:"",themes_branch:main,version:draft'`
   - Browse to http://localhost:8043/admin/websites/website/ and find any site you want to use to testing. Open the admin page for the site and blank out all fields relating to draft and live publish status, then click save
   - Ensure the single site pipeline for said website is pushed up to your local Concourse by running `docker-compose exec web ./manage.py backpopulate_pipelines --filter site-id` where `site-id` is the `name` or `short_id` of your site
   - Browse to the site's editor in the `ocw-studio` UI at http://localhost:8043/sites
   - Publish the site to draft
   - Browse to http://localhost:8080/?search=team%3A%22main%22%20group%3A%22mass-build-sites%22 and after a short while, you should see 2 mass build pipelines appear, both with `version: draft`, one with `offline: false` and the other with `offline: true`
   - Back in the `ocw-studio` UI for the site, wait for the status to change to "Succeded" and then publish the site live
   - Open the Concourse link above again, and after a short while you should see 2 more pipelines appear, this time with `version: live`
 - Browse to the Concourse UI at http://localhost:8080
 - Trigger the live online mass builds and wait for it to complete
 - Trigger the live offline mass builds and wait for it to complete
 - After both builds have finished you should have a fully usable OCW site in your local system. Verify the following:
   - Browse to http://localhost:8045 and verify that you see the home page
   - Click on any of the courses under "new courses" and verify that you see a course site and can browse it normally
   - Browse to the download page and verify that you can download the course ZIP archive
   - Extract the archive and verify that the offline version of the site works
   - Browse to the Minio UI at http://localhost:9001, go to the offline bucket and download one of the folders and ensure you can browse the site offline


## Checklist:
- [ ] After this PR is *released*, a `salt-ops` PR should be put up to add `content_sync.tasks.update_mass_build_pipelines` to `PREPUBLISH_ACTIONS`
